File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int size=0;
        ListNode temp=head;
        while(temp!=null){
            size++;
            temp=temp.next;
        }
        //edge case if first node is the deletion node

        if(n==size){
            return head.next;
        }
        int steps=size-n;
        ListNode curr=head;
        ListNode prev=null;
        while(steps>0){
            prev=curr;
            curr=curr.next;
            steps--;

        }
prev.next=curr.next;
return head;    
   }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length=0;
       
        ListNode* dummynode=new ListNode(0);
        dummynode->next=head;
        ListNode* curr=head;
        while(curr!=NULL){
            length++;
            curr=curr->next;
        }
        curr=dummynode;
        int pos=length-n;
        for(int i=0;i<pos;i++){
           curr=curr->next;
        }
        if(curr->next!=NULL){
            curr->next=curr->next->next;
        }
        return dummynode->next;
    }
};